### DISCUSSION TITLE
Network Speed

### DISCUSSION THREAD

**Discussion by CtheCondor** [RaspAP/raspap-webgui#765] [Q&A]
Is there any software limitation that might slow the speed achieved through the Hot Spot or is that limitation entirely hardware?

--Router Direct--
Ping 6
Download 316 Mbps
Upload 446 Mbps

--RPI AP
Ping 9
Download 25 Mbps
Upload 32 Mbps

--RPI VPN AP
Ping 83
Download 34 Mbps
Upload 8 Mbps

**Comment by billz** [comment]
Short answer: it's both. 

Longer answer: your router is built with application-specific integrated circuits (ASICs) and optimized for computationally heavy but relatively simple networking tasks. The Pi relies on a general-purpose CPU and OS running task-specific software (for example, hostapd) along with hundreds of other services all managed by the kernel. With 5GHz AC wireless and 80MHz-wide channels you can get the RPi4's throughput up to ~75-100Mbps (search issues in this repo for AC mode iperf test results). In an ideal RF environment this can theoretically be higher, but real-world numbers tend to be in this range.

A VPN tunnel adds more latency due to overhead incurred in the process of packet encapsulation. When packets are sent through the tunnel, headers and trailers of various protocols are added to the payload, along with encryption ciphers to secure the tunnel. The RPi4 CPU, while relatively powerful, generally speaking, lacks hardware crypto acceleration. You will, for example, see a difference in throughput with the RC4 cipher (less secure) versus AES (more secure) so there is a tradeoff here.

This is a high level overview. I'm sure others can contribute a more detailed explanation.

**Comment by CtheCondor** [comment]
Thank you for your response and clarification. I was using a dongle that
was advertised as achieving speeds up to 750 Mbps however the driver was
not natively supported so after having to manually search and build for
that driver dependency I didn't know if all that screwed with the
efficiency of the program but then began to question if the software would
affect the efficiency as opposed to just the hardware itself.
